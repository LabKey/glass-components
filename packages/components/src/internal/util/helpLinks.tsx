import React, { FC, ReactNode, memo } from 'react';
import { getServerContext } from '@labkey/api';

export const CHART_MEASURES_AND_DIMENSIONS_TOPIC = 'chartTrouble';
export const MISSING_VALUES_TOPIC = 'manageMissing';
export const PROPERTY_FIELDS_PHI_TOPIC = 'phiLevels';

export const FIELD_EDITOR_TOPIC = 'fieldEditor';
export const ADVANCED_FIELD_EDITOR_TOPIC = FIELD_EDITOR_TOPIC + '#advanced';
export const FIELD_EDITOR_RANGE_VALIDATION_TOPIC = FIELD_EDITOR_TOPIC + '#range';
export const FIELD_EDITOR_REGEX_TOPIC = FIELD_EDITOR_TOPIC + '#regex';
export const FIELD_EDITOR_CONDITIONAL_FORMAT_TOPIC = FIELD_EDITOR_TOPIC + '#conditional';
export const FIELD_EDITOR_SAMPLE_TYPES_TOPIC = FIELD_EDITOR_TOPIC + '#samp';
export const DATE_FORMATS_TOPIC = 'dateFormats#date';
export const NUMBER_FORMATS_TOPIC = 'dateFormats#number';
export const ONTOLOGY_LOOKUP_TOPIC = 'ontologyLookup';
export const ONTOLOGY_CONCEPT_TOPIC = 'ontologyConcept';

export const ASSAY_EDIT_PLATE_TEMPLATE_TOPIC = 'editPlateTemplate';
export const CONFIGURE_SCRIPTING_TOPIC = 'configureScripting';
export const PROGRAMMATIC_QC_TOPIC = 'programmaticQC';
export const DEFINE_ASSAY_SCHEMA_TOPIC = 'defineAssaySchema';
export const DELETE_ASSAY_RUNS_TOPIC = 'manageAssayData#deleteRun';
export const DEFINE_DATA_CLASS_TOPIC = 'createDataClass';
export const DATA_CLASS_NAME_EXPRESSION_TOPIC = 'dataClass#name';
export const DEFINE_SAMPLE_TYPE_TOPIC = 'createSampleSet';
export const DEFINE_LIST_TOPIC = 'createListOptions';
export const DEFINE_ISSUES_LIST_TOPIC = 'adminIssues';
export const DEFINE_DATASET_TOPIC = 'createDataset';
export const DATASET_PROPERTIES_TOPIC = 'datasetProperties#advanced';

export const DELETE_SAMPLES_TOPIC = 'viewSampleSets#delete';
export const DERIVE_SAMPLES_TOPIC = 'deriveSamples';
export const DERIVE_SAMPLES_ALIAS_TOPIC = DERIVE_SAMPLES_TOPIC + '#alias';

export const URL_ENCODING_TOPIC = 'urlEncoding';

export const SEARCH_SYNTAX_TOPIC = 'luceneSearch';
export const DATA_IMPORT_TOPIC = 'dataImport';

export const SAMPLE_ALIQUOT_FIELDS_TOPIC = 'createSampleType#ali';

export const LKS_SAMPLE_ALIQUOT_FIELDS_TOPIC = 'generateSamples#fields';

export const UNIQUE_IDS_TOPIC = 'uniqueStorageIds';

export const CUSTOM_VIEW = 'customViews';

// See HelpTopic.java Referrer enum
export enum HELP_LINK_REFERRER {
    DEV_MENU = 'devMenu',
    DOC_MENU = 'docMenu',
    ERROR_PAGE = 'errorPage',
    IN_PAGE = 'inPage',
    PRODUCT_MENU = 'productMenu',
}

export function getHelpLink(topic: string, referrer = HELP_LINK_REFERRER.IN_PAGE): string {
    // putting referrer= at the end causes links that go to anchors in the page not to work.
    return getServerContext().helpLinkPrefix.replace("name=", "referrer=" + referrer + "&name=" + topic);
}

interface HelpLinkProps {
    className?: string;
    referrer?: HELP_LINK_REFERRER;
    topic: string;
}

export const HelpLink: FC<HelpLinkProps> = props => {
    const { className, topic, referrer, children } = props;

    return (
        <a target="_blank" href={getHelpLink(topic, referrer)} className={className} rel="noopener noreferrer">
            {children}
        </a>
    );
};

/**
 * @deprecated use <HelpLink>
 */
export function helpLinkNode(topic: string, text: ReactNode, className?: string): ReactNode {
    return (
        <HelpLink topic={topic} className={className}>
            {text}
        </HelpLink>
    );
}

interface JavaDocsLinkProps {
    urlSuffix: string;
}

export const JavaDocsLink: FC<JavaDocsLinkProps> = memo(props => {
    const { urlSuffix, children } = props;
    const { jdkJavaDocLinkPrefix } = getServerContext();

    return (
        <a target="_blank" href={jdkJavaDocLinkPrefix + urlSuffix} rel="noopener noreferrer">
            {children}
        </a>
    );
});
