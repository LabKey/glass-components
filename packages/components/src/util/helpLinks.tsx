import React from 'react';

export const CHART_MEASURES_AND_DIMENSIONS_TOPIC = "chartTrouble";
export const MISSING_VALUES_TOPIC = "manageMissing";
export const PROPERTY_FIELDS_PHI_TOPIC = "propertyFields#phi";
export const ADVANCED_PROPERTY_FIELDS_TOPIC =  "propertyFields#advanced";

export const FIELD_EDITOR_TOPIC = "fieldEditor";
export const ADVANCED_FIELD_EDITOR_TOPIC = FIELD_EDITOR_TOPIC + "#advanced";
export const FIELD_EDITOR_RANGE_VALIDATION_TOPIC = FIELD_EDITOR_TOPIC + "#range";
export const FIELD_EDITOR_REGEX_TOPIC = FIELD_EDITOR_TOPIC + "#regex";
export const FIELD_EDITOR_CONDITIONAL_FORMAT_TOPIC = FIELD_EDITOR_TOPIC + "#conditional";
export const FIELD_EDITOR_SAMPLE_TYPES_TOPIC = FIELD_EDITOR_TOPIC + "#samp";
export const DATE_FORMATS_TOPIC = 'dateFormats#date';
export const NUMBER_FORMATS_TOPIC = 'dateFormats#number';

export const ASSAY_EDIT_PLATE_TEMPLATE_TOPIC = 'editPlateTemplate';
export const CONFIGURE_SCRIPTING_TOPIC =  "configureScripting";
export const PROGRAMMATIC_QC_TOPIC = "programmaticQC";
export const DEFINE_ASSAY_SCHEMA_TOPIC = "defineAssaySchema";
export const DEFINE_DATA_CLASS_TOPIC = "dataClass";

export const DELETE_SAMPLES_TOPIC = "viewSampleSets#delete";
export const DERIVE_SAMPLES_TOPIC = 'deriveSamples';
export const DERIVE_SAMPLES_ALIAS_TOPIC = DERIVE_SAMPLES_TOPIC + '#alias';
// export const DERIVE_SAMPLES_GRAPH_TOPIC = DERIVE_SAMPLES_TOPIC + '#graph';
// export const DERIVE_SAMPLES_GRID_TOPIC = DERIVE_SAMPLES_TOPIC + '#grid';

export const URL_ENCODING_TOPIC = 'urlEncoding';

export const SEARCH_SYNTAX_TOPIC = "luceneSearch";
export const DATA_IMPORT_TOPIC = 'dataImport';

export function getHelpLink(topic: string) {
    return LABKEY.helpLinkPrefix + topic;
}

export function helpLinkNode(topic: string, text: React.ReactNode, className?: string): React.ReactNode {
    return (
        <a target="_blank" href={getHelpLink(topic)} className={className}>{text}</a>
    );
}
