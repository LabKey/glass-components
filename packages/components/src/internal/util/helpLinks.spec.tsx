import React from 'react';
import renderer from 'react-test-renderer';

import { HelpLink, getHelpLink, HELP_LINK_REFERRER, JavaDocsLink } from './helpLinks';
import { mount } from 'enzyme';

const HELP_LINK_BASE_URL = 'https://www.labkey.org/Documentation/wiki-page.view?';

describe('HelpLink', () => {
    test('default props', () => {
        const wrapper = mount(<HelpLink topic="TEST_TOPIC">default props text</HelpLink>);
        expect(wrapper.find("a").prop("href")).toBe(HELP_LINK_BASE_URL + 'referrer=inPage&name=TEST_TOPIC');
    });

    test('custom props', () => {
        const wrapper = mount(
            <HelpLink topic="TEST_TOPIC" className="test-class-name" referrer={HELP_LINK_REFERRER.ERROR_PAGE}>
                custom props
            </HelpLink>
        );
        const link = wrapper.find('a');
        expect(link.prop("href")).toBe(HELP_LINK_BASE_URL + 'referrer=errorPage&name=TEST_TOPIC');
        expect(link.prop("className")).toBe('test-class-name');
        expect(wrapper.text()).toBe("custom props");
    });
});

describe('getHelpLink', () => {
    test('default props', () => {
        expect(getHelpLink('TEST_TOPIC')).toBe(HELP_LINK_BASE_URL + 'referrer=inPage&name=TEST_TOPIC');
    });

    test('with referrer', () => {
        expect(getHelpLink('TEST_TOPIC', HELP_LINK_REFERRER.ERROR_PAGE)).toBe(
            HELP_LINK_BASE_URL + 'referrer=errorPage&name=TEST_TOPIC'
        );
    });
});

describe('JavaDocsLink', () => {
    test('default props', () => {
        LABKEY.jdkJavaDocLinkPrefix = 'https://docs.oracle.com/en/java/javase/16/docs/api/java.base/';
        const component = <JavaDocsLink urlSuffix="test/SomeClass.html">Some Class</JavaDocsLink>;
        const tree = renderer.create(component);
        expect(tree).toMatchSnapshot();
    });
});
