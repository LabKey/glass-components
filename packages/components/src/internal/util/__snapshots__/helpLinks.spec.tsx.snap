// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HelpLink custom props 1`] = `
<a
  className="test-class-name"
  href="https://www.labkey.org/Documentation/wiki-page.view?name=TEST_TOPIC&referrer=errorPage"
  rel="noopener noreferrer"
  target="_blank"
>
  custom props
</a>
`;

exports[`HelpLink default props 1`] = `
<a
  href="https://www.labkey.org/Documentation/wiki-page.view?name=TEST_TOPIC&referrer=inPage"
  rel="noopener noreferrer"
  target="_blank"
>
  default props text
</a>
`;

exports[`JavaDocsLink default props 1`] = `
<a
  href="https://docs.oracle.com/en/java/javase/16/docs/api/java.base/test/SomeClass.html"
  rel="noopener noreferrer"
  target="_blank"
>
  Some Class
</a>
`;
